File: scripts/validate-distribution.ts
import { readdir, stat } from 'node:fs/promises';
import { basename, join } from 'node:path';

const repositoryRoot = new URL('../', import.meta.url);
const pluginRoot = new URL('../plugins/feedback-server/', import.meta.url);
const codexManifestPath = new URL('.codex-plugin/plugin.json', pluginRoot);
const codexMarketplacePath = new URL('../.agents/plugins/marketplace.json', import.meta.url);
const claudeManifestPath = new URL('.claude-plugin/plugin.json', pluginRoot);
const claudeMarketplacePath = new URL('../.claude-plugin/marketplace.json', import.meta.url);
const packagePath = new URL('package.json', pluginRoot);
const cursorExamplePath = new URL('../examples/cursor.mcp.json', import.meta.url);
const openCodeExamplePath = new URL('../examples/opencode.json', import.meta.url);

const codexManifest = await Bun.file(codexManifestPath).json() as Record<string, unknown>;
const claudeManifest = await Bun.file(claudeManifestPath).json() as Record<string, unknown>;
const pluginPackage = await Bun.file(packagePath).json() as Record<string, unknown>;
const cursorExample = await Bun.file(cursorExamplePath).json() as Record<string, unknown>;
const openCodeExample = await Bun.file(openCodeExamplePath).json() as Record<string, unknown>;
const codexMarketplace = await Bun.file(codexMarketplacePath).json() as {
  name?: unknown;
  plugins?: Array<{
    name?: unknown;
    source?: { source?: unknown; path?: unknown };
    policy?: { installation?: unknown; authentication?: unknown };
    category?: unknown;
  }>;
};
const claudeMarketplace = await Bun.file(claudeMarketplacePath).json() as {
  name?: unknown;
  plugins?: Array<{
    name?: unknown;
    source?: unknown;
    version?: unknown;
    strict?: unknown;
  }>;
};

function requireCondition(condition: unknown, message: string): asserts condition {
  if (!condition) throw new Error(message);
}

requireCondition(codexManifest.name === basename(pluginRoot.pathname), 'Plugin folder and Codex manifest name differ');
requireCondition(claudeManifest.name === codexManifest.name, 'Agent manifest names differ');
requireCondition(
  typeof codexManifest.version === 'string' && /^\d+\.\d+\.\d+$/.test(codexManifest.version),
  'Plugin version must use strict semver',
);
requireCondition(claudeManifest.version === codexManifest.version, 'Agent manifest versions differ');
requireCondition(pluginPackage.version === codexManifest.version, 'Package and manifest versions differ');
requireCondition(codexManifest.license === 'MIT' && claudeManifest.license === 'MIT', 'Plugin manifests must declare MIT');
requireCondition(codexManifest.skills === './skills/', 'Codex skills path is invalid');
requireCondition(
  typeof codexManifest.mcpServers === 'object'
  && codexManifest.mcpServers !== null
  && 'feedback-server' in codexManifest.mcpServers,
  'Codex MCP configuration is invalid',
);
requireCondition(!('mcpServers' in claudeManifest), 'Claude must use the canonical root .mcp.json');
requireCondition(codexMarketplace.name === 'feedback-server', 'Codex marketplace name is invalid');
requireCondition(codexMarketplace.plugins?.length === 1, 'Codex marketplace must contain exactly one plugin');
const entry = codexMarketplace.plugins[0]!;
requireCondition(entry.name === 'feedback-server', 'Marketplace plugin name is invalid');
requireCondition(entry.source?.source === 'local', 'Marketplace source type is invalid');
requireCondition(entry.source.path === './plugins/feedback-server', 'Marketplace plugin path is invalid');
requireCondition(entry.policy?.installation === 'AVAILABLE', 'Marketplace installation policy is invalid');
requireCondition(entry.policy.authentication === 'ON_INSTALL', 'Marketplace authentication policy is invalid');
requireCondition(entry.category === 'Productivity', 'Marketplace category is invalid');
requireCondition(claudeMarketplace.name === 'feedback-server', 'Claude marketplace name is invalid');
requireCondition(claudeMarketplace.plugins?.length === 1, 'Claude marketplace must contain exactly one plugin');
const claudeEntry = claudeMarketplace.plugins[0]!;
requireCondition(claudeEntry.name === 'feedback-server', 'Claude marketplace plugin name is invalid');
requireCondition(claudeEntry.source === './plugins/feedback-server', 'Claude marketplace source is invalid');
requireCondition(claudeEntry.version === codexManifest.version, 'Claude marketplace version differs');
requireCondition(claudeEntry.strict === true, 'Claude marketplace must use strict plugin metadata');
requireCondition(
  typeof cursorExample.mcpServers === 'object' && cursorExample.mcpServers !== null,
  'Cursor MCP example is invalid',
);
requireCondition(
  typeof openCodeExample.mcp === 'object' && openCodeExample.mcp !== null,
  'OpenCode MCP example is invalid',
);

for (const relativePath of [
  'skills',
  '.claude-plugin/plugin.json',
  '.mcp.json',
  'bin/feedback-server',
  'dist/server.mjs',
  'dist/cli.mjs',
]) {
  try {
    await stat(new URL(relativePath, pluginRoot));
  } catch {
    throw new Error(`Missing plugin component: ${relativePath}`);
  }
}

const allowedRootEntries = new Set([
  '.agents',
  '.claude-plugin',
  '.git',
  '.github',
  '.gitignore',
  'CHANGELOG.md',
  'LICENSE',
  'README.md',
  'docs',
  'examples',
  'package.json',
  'plugins',
  'scripts',
]);
const rootEntries = await readdir(repositoryRoot);
const unexpected = rootEntries.filter((entryName) => !allowedRootEntries.has(entryName));
requireCondition(
  unexpected.length === 0,
  `Unexpected public repository paths: ${unexpected.join(', ')}`,
);

const forbiddenNames = new Set([
  '.env',
  'docker-compose.yml',
  'drizzle.config.ts',
  'DOKPLOY.md',
]);

async function rejectForbiddenPaths(directory: string): Promise<void> {
  for (const entryName of await readdir(directory, { withFileTypes: true })) {
    if (entryName.name === '.git' || entryName.name === 'node_modules') continue;
    const path = join(directory, entryName.name);
    requireCondition(!forbiddenNames.has(entryName.name), `Forbidden public path: ${path}`);
    if (entryName.isDirectory()) await rejectForbiddenPaths(path);
  }
}

await rejectForbiddenPaths(repositoryRoot.pathname);
console.error(`FeedbackServer Plugin ${String(codexManifest.version)} is structurally valid for Codex and Claude Code.`);

File: scripts/validate_distribution.py
#!/usr/bin/env python3
"""Validate the runtime-free FeedbackKit 2.0 remote MCP plugin distribution."""

from __future__ import annotations

import json
import re
from pathlib import Path


ROOT = Path(__file__).resolve().parent.parent
PLUGIN = ROOT / "plugins" / "feedback-server"
VERSION = "2.0.0"
MCP_URL = "https://api.feedkit.cn/mcp"
SKILLS = (
    "setup-feedback-server",
    "triage-feedback",
    "manage-waitlist",
    "publish-roadmap-releases",
    "configure-notifications",
    "administer-feedback-server",
    "manage-feedback-server",
)
EVALS = (
    "browser-oauth",
    "error-recovery",
    "negative-unrelated-feedback",
    "publishing",
    "setup",
    "triage",
    "triage-direct-write",
    "waitlist",
)


def load(path: Path) -> dict:
    with path.open(encoding="utf-8") as handle:
        value = json.load(handle)
    require(isinstance(value, dict), f"{path} must contain a JSON object")
    return value


def require(condition: bool, message: str) -> None:
    if not condition:
        raise RuntimeError(message)


def text(path: Path) -> str:
    return path.read_text(encoding="utf-8")


package = load(ROOT / "package.json")
codex = load(PLUGIN / ".codex-plugin" / "plugin.json")
claude = load(PLUGIN / ".claude-plugin" / "plugin.json")
mcp = load(PLUGIN / ".mcp.json")
codex_marketplace = load(ROOT / ".agents" / "plugins" / "marketplace.json")
claude_marketplace = load(ROOT / ".claude-plugin" / "marketplace.json")
cursor = load(ROOT / "examples" / "cursor.mcp.json")
opencode = load(ROOT / "examples" / "opencode.json")

require(package.get("version") == VERSION, f"Package distribution version must be {VERSION}")
require(codex.get("name") == PLUGIN.name, "Plugin folder and Codex manifest name differ")
require(codex.get("version") == VERSION, f"Codex plugin version must be {VERSION}")
require(claude.get("version") == VERSION, f"Claude plugin version must be {VERSION}")
require(codex.get("skills") == "./skills/", "Codex skills path is invalid")
require(
    codex.get("mcpServers") == {
        "feedback-server": {"type": "http", "url": MCP_URL},
    },
    "Codex must declare the remote HTTP MCP endpoint",
)
require("apps" not in codex, "Do not add an app manifest until the real connector ID is registered")
require("mcpServers" not in claude, "Claude must use the canonical root .mcp.json")
require(
    mcp.get("feedback-server") == {"type": "http", "url": MCP_URL},
    "Remote MCP descriptor is invalid",
)

codex_entries = codex_marketplace.get("plugins")
require(isinstance(codex_entries, list) and len(codex_entries) == 1, "Codex marketplace must contain one plugin")
codex_entry = codex_entries[0]
require(codex_entry.get("name") == "feedback-server", "Codex marketplace plugin name is invalid")
require(
    codex_entry.get("source") == {"source": "local", "path": "./plugins/feedback-server"},
    "Codex marketplace source is invalid",
)
require(
    codex_entry.get("policy") == {"installation": "AVAILABLE", "authentication": "ON_USE"},
    "Codex marketplace must defer browser OAuth until first protected use",
)

claude_entries = claude_marketplace.get("plugins")
require(isinstance(claude_entries, list) and len(claude_entries) == 1, "Claude marketplace must contain one plugin")
claude_entry = claude_entries[0]
require(claude_entry.get("version") == VERSION, "Claude marketplace version differs")
require(claude_entry.get("strict") is True, "Claude marketplace must use strict metadata")

require(
    cursor.get("mcpServers", {}).get("feedback-server", {}).get("url") == MCP_URL,
    "Cursor example is not remote",
)
require(
    opencode.get("mcp", {}).get("feedback-server", {}).get("url") == MCP_URL,
    "OpenCode example is not remote",
)

declared_claude_skills = claude.get("skills")
require(
    declared_claude_skills == [f"./skills/{skill}" for skill in SKILLS],
    "Claude manifest skill list is incomplete or out of order",
)

skill_texts: dict[str, str] = {}
for skill in SKILLS:
    path = PLUGIN / "skills" / skill / "SKILL.md"
    require(path.is_file(), f"Missing skill: {skill}")
    content = text(path)
    skill_texts[skill] = content
    require("2.0" in content, f"Skill {skill} must explicitly target FeedbackServer 2.0")
    require("elicitation" in content, f"Skill {skill} must prefer MCP elicitation for risky tools")
    require("`confirm: true`" in content, f"Skill {skill} must document the v2 fallback confirmation")
    metadata = text(PLUGIN / "skills" / skill / "agents" / "openai.yaml")
    require(
        'transport: "streamable_http"' in metadata,
        f"Skill {skill} must declare the Streamable HTTP MCP transport",
    )

setup_skill = skill_texts["setup-feedback-server"]
for required in (
    "Better Auth",
    "magic link",
    "`whoami`",
    "`list_products`",
    "`create_product`",
    "`set_notification_setup_preference`",
    "`mcp_server`",
    "`account_email`",
):
    require(required in setup_skill, f"Setup skill is missing the v2 setup contract: {required}")

triage_skill = skill_texts["triage-feedback"]
normalized_triage_skill = " ".join(triage_skill.split())
for required in ("`update_feedback`", "`reply_to_feedback`", "without asking the user to approve the same action again"):
    require(required in normalized_triage_skill, f"Triage skill is missing direct-write intent: {required}")

root_readme = text(ROOT / "README.md")
for required in ("FeedbackKit Agent Plugin 2.0", "Better Auth", "`whoami`", "`confirm: true`"):
    require(required in root_readme, f"Root README is missing the v2 contract: {required}")

invitation_docs = (
    ROOT / "docs" / "invited-admin-onboarding.en.md",
    ROOT / "docs" / "invited-admin-onboarding.zh-Hans.md",
)
instruction_field = re.compile(r"^\s*([a-z][a-z0-9_]*):\s+", re.MULTILINE)
for path in invitation_docs:
    content = text(path)
    fields = set(instruction_field.findall(content))
    require(fields == {"mcp_server", "account_email"}, f"Invitation instructions in {path.name} must use only mcp_server and account_email")
    require("`whoami`" in content and "magic" in content.casefold(), f"Invitation instructions in {path.name} must use the v2 browser flow")

eval_root = PLUGIN / "evals"
actual_evals = tuple(sorted(path.name for path in eval_root.iterdir() if path.is_dir()))
require(actual_evals == EVALS, "Eval directories must exactly cover the FeedbackServer 2.0 suite")
for eval_name in EVALS:
    path = eval_root / eval_name / "case.yaml"
    require(path.is_file(), f"Missing eval case: {eval_name}")
    content = text(path)
    for marker in (
        'schema_version: "1.1"',
        "description:",
        "execution:",
        "  prompt:",
        "  allowed_tools:",
        "graders:",
        "expected_outcome:",
    ):
        require(marker in content, f"Eval {eval_name} is missing required field: {marker.strip()}")
    require("v2" in content.casefold() or "2.0" in content, f"Eval {eval_name} does not explicitly target v2")

contract_paths = sorted(
    path
    for path in ROOT.rglob("*")
    if path.is_file()
    and ".git" not in path.parts
    and "__pycache__" not in path.parts
)
contract_contents: list[tuple[Path, str]] = []
for path in contract_paths:
    try:
        content = text(path)
    except UnicodeDecodeError:
        continue
    contract_contents.append((path, content))

retired_tool_names = (
    "authen" + "ticate",
    "authentication_" + "status",
    "connection_" + "status",
    "get_onboarding_" + "status",
    "execute_" + "confirmation",
)
for path, content in contract_contents:
    for tool_name in retired_tool_names:
        pattern = re.compile(rf"(?<![A-Za-z0-9_]){re.escape(tool_name)}(?![A-Za-z0-9_])", re.IGNORECASE)
        require(not pattern.search(content), f"Retired tool name remains in {path.relative_to(ROOT)}")

retired_setup_markers = (
    "pending_" + "verification",
    "empty " + "scopes list",
    "unbound " + "connection",
    "account " + "binding",
    "invitation " + "code",
    "short-" + "code",
    "pairing " + "code",
    "verification " + "code",
    'method: "' + 'invitation"',
    'method: "' + 'email"',
)
for path, content in contract_contents:
    folded = content.casefold()
    found = [marker for marker in retired_setup_markers if marker.casefold() in folded]
    require(not found, f"Retired setup guidance remains in {path.relative_to(ROOT)}")

for forbidden in ("bin", "dist", "src", "scripts", "tests"):
    directory = PLUGIN / forbidden
    require(
        not directory.exists() or not any(path.is_file() for path in directory.rglob("*")),
        f"Runtime directory must not ship files: {forbidden}",
    )
for forbidden in ("bun.lock", "package.json", "tsconfig.json", "eslint.config.js"):
    require(not (PLUGIN / forbidden).exists(), f"Runtime file must not ship: {forbidden}")

allowed_root = {
    ".agents", ".claude-plugin", ".git", ".github", ".gitignore", "CHANGELOG.md",
    "LICENSE", "README.md", "docs", "examples", "package.json", "plugins", "scripts",
}
unexpected = sorted(path.name for path in ROOT.iterdir() if path.name not in allowed_root)
require(not unexpected, f"Unexpected public repository paths: {', '.join(unexpected)}")

private_names = {".env", "docker-compose.yml", "drizzle.config.ts", "DOKPLOY.md"}
private_paths = sorted(
    str(path.relative_to(ROOT))
    for path in ROOT.rglob("*")
    if path.is_file() and ".git" not in path.parts and path.name in private_names
)
require(
    not private_paths,
    f"Private deployment files must not ship: {', '.join(private_paths)}",
)

print(f"FeedbackServer Plugin {VERSION} is a valid v2-only runtime-free remote MCP distribution.")
